### PR TITLE
fix: mypy issue on py3.9 + prevent similar issues

### PR DESCRIPTION
**Pull request by mistercrunch**
Someone reported that they had this mypy-related issue on py3.9, so I decided to fix it and to run pre-commit checks against more python versions to prevent this kind of issue from slipping into the repo

### CODE REVIEW

**Comment by rusackas** [comment]
Not sure why the pre-commit job canceled here.

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -16,6 +16,9 @@ concurrency:
 jobs:
   pre-commit:
     runs-on: ubuntu-22.04
+    strategy:
+      matrix:
+        python-version: ["current", "next", "previous"]
```
Comment (rusackas) [inline .github/workflows/pre-commit.yml:21]: Should we make this `python-version: ["3.9", "3.10", "3.11"]` to be a little more explicit in what we support?

Comment (mistercrunch) [reply .github/workflows/pre-commit.yml:21]: It's like this for very good reason - has to do with `required_status_checks` and matrix-management. There's all sorts of issues around renaming checks/matrix members and required checks. When we'll bump python version we'll be able to do it centrally in `.github/actions/setup-backend/action.yml` (current/next/future is defined by us in there), and we won't have issues around having to rename check names in master prior to being able to merge the branch

Comment (mistercrunch) [reply .github/workflows/pre-commit.yml:21]: This recent PR shows the kind of trouble we have to go through in order to rename checks... https://github.com/apache/superset/pull/29866

File: superset/commands/distributed_lock/base.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
 
 import logging
 import uuid
-from typing import Any
+from typing import Any, Union
```
Comment (mistercrunch) [inline superset/commands/distributed_lock/base.py:20]: python 3.9 doesn't support the pipe operator in type hints...

**Review by rusackas — APPROVED** [review]

**Pull request merged**